x-pack/plugins/synthetics/public/apps/synthetics/components/monitor_details/monitor_selector/use_recently_viewed_monitors.ts

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import useLocalStorage from 'react-use/lib/useLocalStorage';
+import { i18n } from '@kbn/i18n';
+import { useParams } from 'react-router-dom';
+import { isEqual } from 'lodash';
+import { useEffect, useMemo } from 'react';
+import { useKibana } from '@kbn/kibana-react-plugin/public';
+import { useFetcher } from '@kbn/observability-plugin/public';
+import { MonitorFields } from '../../../../../../common/runtime_types';
+import { syntheticsMonitorType } from '../../../../../../common/types/saved_objects';
+
+export const useRecentlyViewedMonitors = () => {
+  const [recentlyViewed, setRecentlyViewed] = useLocalStorage<string[]>(
+    'xpack.synthetics.recentlyViewedMonitors',
+    []
+  );
+  const { monitorId } = useParams<{ monitorId: string }>();
+
+  const { savedObjects } = useKibana().services;
+
+  useEffect(() => {
+    const newRecentlyViewed = [
+      ...new Set([...(monitorId ? [monitorId] : []), ...(recentlyViewed ?? [])]),
+    ].slice(0, 5);
+
+    if (!isEqual(newRecentlyViewed.sort(), recentlyViewed?.sort())) {
```

Comment:
If `monitorId` changes, we should always update the cached value, to let the newly visited monitor to go on top.

If without this, updates are happening excessively frequent, try to optimize from the top of component tree. A check here could be placed here if `monitorId` switches from a null/empty value to a valid value or vice versa.

Bug's scenario, recently visited monitor should go to the top:

https://user-images.githubusercontent.com/2748376/194903197-bffdcd79-5520-4c66-b885-7d0e2c37af48.mov